src/Compilers/CSharp/Portable/Binder/Binder.ValueChecks.cs

```diff
@@ -894,120 +936,345 @@ private bool CheckPropertyValueKind(SyntaxNode node, BoundExpression expr, BindV
+            //                                                its val escape is ExternalScope                   (does not affect overal result)
+            if (unmatchedInParameter != null)
+            {
+                Error(diagnostics, GetStandardCallEscapeError(checkingReceiver), syntax, symbol, unmatchedInParameter.Name);
+                return false;
             }
 
-            // check all "in" parameters 
-            for (var paramIndex = 0; paramIndex < parameters.Length; paramIndex++)
+            // check receiver if ref-like
+            if (receiverOpt?.Type?.IsByRefLikeType == true)
             {
-                var parameter = parameters[paramIndex];
+                return CheckValEscape(receiverOpt.Syntax, receiverOpt, escapeFrom, escapeTo, false, diagnostics);
+            }
 
-                if (parameter.RefKind != RefKind.RefReadOnly)
-                {
-                    continue;
-                }
+            return true;
+        }
 
-                if (parameter.IsParams)
-                {
-                    break;
-                }
+        /// <summary>
+        /// Validates whether the invocation is valid per no-mixing rules.
+        /// Returns `false` when it is not valid and produces diagnostics (possibly more than one recursively) that helps to figure the reason.
+        /// </summary>
+        private static bool CheckInvocationArgMixing(
+            SyntaxNode syntax,
+            Symbol symbol,
+            BoundExpression receiverOpt,
+            ImmutableArray<ParameterSymbol> parameters,
+            ImmutableArray<BoundExpression> args,
+            ImmutableArray<RefKind> argRefKinds,
+            ImmutableArray<int> argsToParamsOpt,
+            uint scopeOfTheContainingExpression,
+            DiagnosticBag diagnostics)
+        {
+            // widest possible escape via writeble ref-like receiver or ref/out argument.
+            uint escapeTo = scopeOfTheContainingExpression;
 
-                BoundExpression argument = null;
-                if (argToParamsOpt.IsDefault)
+            // collect all writeable ref-like arguments, including receiver
+            if (receiverOpt?.Type?.IsByRefLikeType == true)
+            {
+                escapeTo = GetValEscape(receiverOpt, scopeOfTheContainingExpression);
+            }
+
+            if (!args.IsDefault)
+            {
+                for (var argIndex = 0; argIndex < args.Length; argIndex++)
                 {
-                    if (paramIndex < args.Length)
+                    var argument = args[argIndex];
+                    var refKind = argRefKinds.IsDefault ? RefKind.None : argRefKinds[argIndex];
+                    if (refKind != RefKind.None && argument.Type?.IsByRefLikeType == true)
                     {
-                        argument = args[paramIndex];
+                        Debug.Assert(refKind == RefKind.Ref || refKind == RefKind.None);
```

Comment:
`refKind` is not `RefKind.None` given the check above.

Reply:
should be Out, Thanks!!!

---
In reply to: [135280866](https://github.com/dotnet/roslyn/pull/21334#discussion_r135280866) [](ancestors = 135280866)